Add a FAQ page to the docs (#2933)
zalando/patroni: docs/faq.rst

What occurs if I change a Postgres GUC that requires a restart?
    Patroni will mark the affected members with a flag of ``pending restart``.

    It is up to you to determine when and how to restart the members. That can be accomplished either through:

    * :ref:`patronictl_restart`; or
    * A ``POST`` request to :ref:`restart_endpoint`.

What is the difference between ``etcd`` and ``etcd3`` in Patroni configuration?
    ``etcd`` uses the API version 2 of ``etcd``, while ``etcd3`` uses the API version 3 of ``etcd``.

    Be aware that information stored by the API version 2 is not manageable by API version 3 and vice-versa.

    The recommended one is ``etcd3`` taking into consideration that ``etcd`` API version 2 end of life was in 2017.

I have ``use_slots`` enabled in my Patroni configuration, but when a cluster member goes offline for some time, the replication slot used by that member is dropped on the upstream node. What can I do to avoid that issue?
    You can configure a permanent physical replication slot for the members.

>     Since Patroni ``3.2.0`` it is now possible to have member slots as permanent slots managed by Patroni.

Reviewer: This is not entirely true. Permanent physical slots that match with member names was possible to have to a few years, but they existed only on the leader. That is, in case of failover/switchover these slots were created on the new leader, but that didn't guaranty that it had all WAL segments for the absent node.

Author: Added a commit with this info.

>     Patroni will create the permanent physical slots on all nodes, and make sure to not remove the slots, as well as to advance the slots' LSN on all nodes according to the LSN that has been consumed by the member.

Reviewer: There is a small race condition. In the very beginning, when the slot is created on the replica it could be ahead of the same slot on the leader and in case if nobody is consuming from the slot there is still a chance that some files could be missing after failover. But in any case this is a good improvement comparing to the old state.

P.S., maybe also recommend to run continues archiving / PITR?

Author: Added a commit with this info.

    You can find more information about that in :ref:`pause`.

Automatic failover
------------------

How does the automatic failover mechanism of Patroni work?
    Patroni automatic failover is based on what we call ``leader race``.

    Patroni stores the cluster's status in the DCS, among them a ``leader`` lock which holds the name of the Patroni member which is the current ``leader`` of the cluster.

    That ``leader`` lock has a time-to-live associated with it. If the leader node fails to update the lease of the ``leader`` lock in time, the key will eventually expire from the DCS.

    When the ``leader`` lock expires, it triggers what Patroni calls a ``leader race``: all nodes start performing checks to determine if they are the best candidates for taking over the ``leader`` role.
    Some of these checks include calls to the REST API of all other Patroni members.

    All Patroni members that find themselves as the best candidate for taking over the ``leader`` lock will attempt to do so.

>     The first Patroni member that is able to take the ``leader`` lock will promote itself to a read/write node, and the others will be configured to follow it.

Reviewer: ```suggestion
    The first Patroni member that is able to take the ``leader`` lock will promote itself to a read/write node (or ``standby leader``), and the others will be configured to follow it.
```

    As the member is not able to update the information, the information eventually expires from the DCS, and consequently the member is not shown anymore in the output of :ref:`patronictl_list`.

Why is the information about one of my Patroni members not up-to-date in the output of :ref:`patronictl_list` command?
    Information shown by :ref:`patronictl_list` is based on the contents of the DCS.

    By default, that information is updated by Patroni roughly every ``loop_wait`` seconds.
    In other words, even if everything is normally functional you may still see a "delay" of up to ``loop_wait`` seconds in the information stored in the DCS.

    Be aware that that is not a rule, though. Some operations performed by Patroni cause it to immediately update the DCS information.

Configuration
-------------

What is the difference between dynamic configuration and local configuration?
    Dynamic configuration (or global configuration) is the configuration stored in the DCS, and which is applied to all members of the Patroni cluster.
    This is primarily where you should store your configuration.

>     Settings that are specific to a node, or settings that you would like the global configuration overwritten, you should set only on the desired Patroni member as a local configuration.

Reviewer 2: ```suggestion
    Settings that are specific to a node, or settings that you would like to overwrite the global configuration with, you should set only on the desired Patroni member as a local configuration.
```
but maybe I don't get the idea of the sentence

Author: That's it, yours looks better :)

    If you started Patroni through other means, you will need to identify the ``patroni`` process and run ``kill -s HUP PID``, ``PID`` being the process ID of the ``patroni`` process.

How can I change my environment configuration?
    The environment configuration is only read by Patroni during startup.

    With that in mind, if you change the environment configuration you will need to restart the corresponding Patroni agent.

    Take care to not cause a failover in the cluster! You might be interested in checking :ref:`patronictl_pause`.

What occurs if I change a Postgres GUC that requires a reload?
    When you change the dynamic or the local configuration as explained in the previous questions, Patroni will take care of reloading the Postgres configuration for you.

What occurs if I change a Postgres GUC that requires a restart?
    Patroni will mark the affected members with a flag of ``pending restart``.

    It is up to you to determine when and how to restart the members. That can be accomplished either through:

    * :ref:`patronictl_restart`; or

>     * A ``POST`` request to :ref:`restart_endpoint`.

Author: Cool, good point, I'll add a reference to the relevant docs that explain that :)

    Later, if you decide to remove the corresponding member, it's **your responsability** to adjust the permanent slots configuration, otherwise Patroni will keep the slots around forever.

    **Note:** on Patroni older than ``3.2.0`` you could still have member slots configured as permanent physical slots, however they would be managed only on the current leader. That is, in case of failover/switchover these slots would be created on the new leader, but that wouldn't guarantee that it had all WAL segments for the absent node.

    **Note:** even with Patroni ``3.2.0`` there might be a small race condition. In the very beginning, when the slot is created on the replica it could be ahead of the same slot on the leader and in case if nobody is consuming the slot there is still a chance that some files could be missing after failover. With that in mind, it is recommended that you configure continuous archiving, which makes it possible to restore required WALs or perform PITR.

What is the difference between ``loop_wait``, ``retry_timeout`` and ``ttl``?

>     Patroni performs what we call a HA cycle from time to time. On each HA cycle it takes care of performing a series of checks on the cluster to determine its healthiness, and depending on the status it may take actions, like failing over to a standby.

Author: Right, I'll add a reference to the corresponding docs page.

>     ``loop_wait`` determines for long, in seconds, Patroni should sleep before performing a new cycle of HA checks.

Reviewer 2: ```suggestion
    ``loop_wait`` determines for how long, in seconds, Patroni should sleep before performing a new cycle of HA checks.
```

    ``retry_timeout`` sets the timeout for retry operations on the DCS and on Postgres. For example: if the DCS is unresponsive for more than ``retry_timeout`` seconds, Patroni might demote the primary node as a security action.

    ``ttl`` sets the lease time on the ``leader`` lock in the DCS. If the current leader of the cluster is not able to renew the lease during its HA cycles for longer than ``ttl``, then the lease will expire and that will trigger a ``leader race`` in the cluster.

> Postgres management

Reviewer: Please also add an item that postgresql.service **MUST** be disabled.
And explanation: if the node running as a primary failed, the postgresql.service shouldn't start postgres automatically because it may cause a split-brain. Only Patroni is allowed to start/stop/promote Postgres.

-------------------

Can I change Postgres GUCs directly in Postgres configuration?
    You should **not** do that!

    Postgres configuration is managed by Patroni, and attempts to edit the configuration files will end up being frustrated by Patroni as it will eventually overwrite them.

>     You need to manage all the Postgres configuration through Patroni!

Author: Thanks, added that info and a link to the "important rules" section.
